smart-contracts/contracts/Unlock.sol

```diff
@@ -126,7 +148,55 @@ contract Unlock is
     initializer()
   {
     // We must manually initialize Ownable
-    Ownable.__initializeOwnable(_unlockOwner);
+    UnlockOwnable.__initializeOwnable(_unlockOwner);
+  }
+
+  function initializeProxyAdmin() public {
+    require(proxyAdminAddress == address(0), "ProxyAdmin already deployed");
+    _deployProxyAdmin();
+  }
+
+  /**
+  * @dev Deploy the ProxyAdmin contract that will manage lock templates upgrades
+  * This deploys an instance of ProxyAdmin used by PublicLock transparent proxies.
+  */
+  function _deployProxyAdmin() private returns(address) {
+    proxyAdmin = new ProxyAdmin();
+    proxyAdminAddress = address(proxyAdmin);
+    return address(proxyAdmin);
+  }
+
+  /**
+  * @dev Helper to get the version number of a template from his address
+  */
+  function publicLockVersions(address _impl) external view returns(uint16) {
+    return _publicLockVersions[_impl];
+  }
+
+  /**
+  * @dev Helper to get the address of a template based on its version number
+  */
+  function publicLockImpls(uint16 _version) external view returns(address) {
+    return _publicLockImpls[_version];
+  }
+
+  /**
+  * @dev Registers a new PublicLock template immplementation
+  * The template is identified by a version number
+  * Once registered, the template can be used to upgrade an existing Lock
+  */
+  function addLockTemplate(address impl, uint16 version) public onlyOwner {
+    require(impl != address(0), "impl address can not be 0x");
+    require(version != 0, "version can not be 0");
+    require(_publicLockVersions[impl] == 0, "address already used by another version");
```

Comment:
I wonder if we should let the admin "overwrite" a function. Maybe it's fine since it's an admin thing anyway?

Reply:
not sure what do you mean by overwrite?

Reply:
Well let the owner (re)set an existing version...